src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs

```diff
@@ -508,6 +517,57 @@ private void ValidateLicenseFile(IEnumerable<IPackageFile> files, LicenseMetadat
             }
         }
 
+        /// <summary>
+        /// Given a list of resolved files,
+        /// determine which file will be used as the icon file and validate its size.
+        /// </summary>
+        /// <param name="files">Files resolved from the file entries in the nuspec</param>
+        /// <param name="iconPath">iconpath found in the .nuspec</param>
+        /// <exception cref="PackagingException">When a validation rule is not met</exception>
+        private void ValidateIconFile(IEnumerable<IPackageFile> files, string iconPath)
+        {
+            if (!string.IsNullOrEmpty(iconPath))
+            {
+                // Validate entry
+                var iconPathStripped = PathUtility.StripLeadingDirectorySeparators(iconPath);
+
+                var iconFileList = files.Where(f =>
+                        iconPath.Equals(
+                            PathUtility.StripLeadingDirectorySeparators(f.Path),
+                            PathUtility.GetStringComparisonBasedOnOS()));
+
+                if (iconFileList.Count() > 1)
+                {
+                    throw new PackagingException(
```

Comment:
As mentioned in the actual bug, this is not icon specific, so I wouldn't handle for the icons only. We should fix the original bug and handle it everywhere.